clients/drcachesim/tests/invariant_checker_test.cpp

```diff
@@ -0,0 +1,331 @@
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL GOOGLE, LLC OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* Test for checks performed by invariant_checker_t that are not tested
+ * by the signal_invariants app's prefetch and handler markers.
+ * This looks for precise error strings from invariant_checker.cpp: but
+ * we will notice if the literals get out of sync as the test will fail.
+ */
+
+#include <iostream>
+#include <vector>
+
+#include "../tools/invariant_checker.h"
+#include "../common/memref.h"
+
+namespace {
+
+class checker_no_abort_t : public invariant_checker_t {
+public:
+    checker_no_abort_t(bool offline)
+        : invariant_checker_t(offline)
+    {
+    }
+    std::vector<std::string> errors;
+
+protected:
+    void
+    report_if_false(bool condition, const std::string &message) override
+    {
+        if (!condition)
+            errors.push_back(message);
+    }
+};
+
+memref_t
+gen_instr_type(trace_type_t type, memref_tid_t tid, addr_t pc)
+{
+    memref_t memref = {};
+    memref.instr.type = type;
+    memref.instr.tid = tid;
+    memref.instr.addr = pc;
+    memref.instr.size = 1;
+    return memref;
+}
+
+memref_t
+gen_instr(memref_tid_t tid, addr_t pc)
+{
+    return gen_instr_type(TRACE_TYPE_INSTR, tid, pc);
+}
+
+memref_t
+gen_branch(memref_tid_t tid, addr_t pc)
+{
+    return gen_instr_type(TRACE_TYPE_INSTR_CONDITIONAL_JUMP, tid, pc);
+}
+
+memref_t
+gen_marker(memref_tid_t tid, trace_marker_type_t type, uintptr_t val)
+{
+    memref_t memref = {};
+    memref.marker.type = TRACE_TYPE_MARKER;
+    memref.marker.tid = tid;
+    memref.marker.marker_type = type;
+    memref.marker.marker_value = val;
+    return memref;
+}
+
+memref_t
+gen_exit(memref_tid_t tid)
+{
+    memref_t memref = {};
+    memref.instr.type = TRACE_TYPE_THREAD_EXIT;
+    memref.instr.tid = tid;
+    return memref;
+}
+
+bool
+check_branch_target_after_branch()
+{
+    // Positive simple test.
+    {
+        checker_no_abort_t checker(true /*offline*/);
+        std::vector<memref_t> memrefs = {
+            gen_instr(1, 1), gen_branch(1, 2),
+            gen_instr(1, 3), gen_marker(2, TRACE_MARKER_TYPE_TIMESTAMP, 0),
```

Comment:
So here the branch is not taken (as the next instruction has pc = prev_pc + 1). Can we add a positive test for a taken branch, where the next instruction pc is the target of the branch?